src/main/kotlin/parts/wisdom/arcraider/visualization/GridPane.kt

```diff
@@ -3,45 +3,64 @@ package parts.wisdom.arcraider.visualization
 import java.awt.Color
 import java.awt.Graphics
 import javax.swing.JPanel
+import kotlin.math.min
 
 // When calculating the width / height of each grid pixel, offset it so
 // we don't fill right to the edge of the screen.
 private const val PANEL_INSET_PIXELS = 10
 
-class GridPane(var grid: Grid) : JPanel() {
-
-    private fun xPixelsPerSquare(): Int = (width - PANEL_INSET_PIXELS) / grid.widthSquares
-    private fun yPixelsPerSquare(): Int = (height - PANEL_INSET_PIXELS) / grid.heightSquares
-    private fun xStartPixels(): Int = (width - xPixelsPerSquare() * grid.widthSquares) / 2
-    private fun yStartPixels(): Int = (height - yPixelsPerSquare() * grid.heightSquares) / 2
+class GridPane(private var grid: Grid, private val isOutput: Boolean = false) : JPanel() {
 
     override fun paintComponent(g: Graphics) {
         super.paintComponent(g)
-        background = Color.BLACK
         val graphics2d = g.create()
-        graphics2d.color = Color.WHITE
 
+        // Draw the background
+        graphics2d.color = Color.BLACK
+        val (xPixels, yPixels) = gridSquaresToPixels(grid.widthSquares, grid.heightSquares)
+
+        val xOffset = if (isOutput) width - xPixels - PANEL_INSET_PIXELS else 0
```

Comment:
Perhaps add a comment to explain that this is to right-justify the output grid? It took me embarrassingly long to figure that out.